src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -0,0 +1,412 @@
+      case _ =>
+        None
+    }
+  }
+
+  private[this] def build(
+    podDefinition: PodDefinition,
+    offer: mesos.Offer,
+    newInstanceId: PathId => Instance.Id,
+    config: MarathonConf,
+    resourceMatch: ResourceMatcher.ResourceMatch
+  ): Some[(mesos.ExecutorInfo, mesos.TaskGroupInfo, Seq[Option[Int]])] = {
+    // TODO: probably set unique ID for each task
+    val instanceId = newInstanceId(podDefinition.id)
+
+    val executorInfo = computeExecutorInfo(podDefinition, instanceId)
+
+    val taskGroup = mesos.TaskGroupInfo.newBuilder
+
+    podDefinition.containers
+      .map(computeTaskInfo(_, podDefinition, offer, instanceId, config, resourceMatch))
+      .foreach(taskGroup.addTasks)
+
+    Some((executorInfo.build, taskGroup.build, resourceMatch.hostPorts))
+  }
+
+  private[this] def computeTaskInfo(
+    container: raml.MesosContainer,
+    podDefinition: PodDefinition,
+    offer: mesos.Offer,
+    instanceId: Instance.Id,
+    config: MarathonConf,
+    resourceMatch: ResourceMatcher.ResourceMatch
+  ): mesos.TaskInfo.Builder = {
+    val builder = mesos.TaskInfo.newBuilder
+      .setName(container.name)
+      .setTaskId(instanceId.mesosTaskId)
+      .setSlaveId(offer.getSlaveId)
+
+    builder.addResources(scalarResource("cpus", container.resources.cpus))
+    builder.addResources(scalarResource("mem", container.resources.mem))
+    builder.addResources(scalarResource("disk", container.resources.disk))
+    builder.addResources(scalarResource("gpus", container.resources.gpus.toDouble))
+
+    val labels = podDefinition.labels ++ container.labels.map(_.values).getOrElse(Map.empty[String, String])
+
+    if (labels.nonEmpty)
+      builder.setLabels(mesos.Labels.newBuilder.addAllLabels(labels.map {
+        case (key, value) =>
+          mesos.Label.newBuilder.setKey(key).setValue(value).build
+      }.asJava))
+
+    val envPrefix: Option[String] = config.envVarsPrefix.get
+
+    val commandInfo = computeCommandInfo(
+      podDefinition,
+      instanceId,
+      container,
+      offer.getHostname,
+      resourceMatch.hostPorts,
+      envPrefix)
+
+    builder.setCommand(commandInfo)
+
+    val containerInfo = computeContainerInfo(podDefinition.podVolumes, container)
+
+    builder.setContainer(containerInfo)
+
+    container.healthCheck.foreach { healthCheck =>
+      builder.setHealthCheck(computeHealthCheck(healthCheck, container.endpoints))
+    }
+
+    builder
+  }
+
+  private[this] def computeExecutorInfo(
+    podDefinition: PodDefinition,
+    instanceId: Instance.Id): mesos.ExecutorInfo.Builder = {
+    val executorID = mesos.ExecutorID.newBuilder.setValue(f"marathon-${instanceId.idString}")
+
+    val executorInfo = mesos.ExecutorInfo.newBuilder
+      .setType(mesos.ExecutorInfo.Type.DEFAULT)
+      .setExecutorId(executorID)
+
+    // We're deliberately not setting executor resources here. Because we're using the default executor, the
```

Comment:
Who sets the executor cpu, mem? Mesos? If so .. wouldn't it be better to be explicit here? We already have constants defined that are taken into consideration during resource matching. Why not reference them here?